sdk/client-core/core/src/main/java/io/clientcore/core/util/ServerSentEventUtils.java

```diff
@@ -56,22 +55,50 @@ public static boolean isTextEventStreamContentType(String contentType) {
      * <p>
      * The passed {@link InputStream} will be closed by this method.
      *
-     * @param httpClient The {@link HttpClient} to send the {@code httpRequest} through.
-     * @param httpRequest The {@link HttpRequest} to send.
      * @param inputStream The {@link InputStream} to read data from.
      * @param listener The {@link ServerSentEventListener event listener} attached to the {@link HttpRequest}.
+     * @return The non-null {@link ServerSentResult result} in case of an error.
      * @throws IOException If an error occurs while reading the data.
      */
-    public static void processTextEventStream(HttpClient httpClient, HttpRequest httpRequest,
-        InputStream inputStream, ServerSentEventListener listener) throws IOException {
+    public static ServerSentResult processTextEventStream(InputStream inputStream,
+                                                          ServerSentEventListener listener) throws IOException {
         try (BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream, StandardCharsets.UTF_8))) {
-            RetrySSEResult retrySSEResult = processBuffer(reader, listener);
+            return processBuffer(reader, listener);
+        }
+    }
+
+    /**
+     * Retries the {@link HttpRequest} if the listener allows it.
+     *
+     * @param serverSentResult The {@link ServerSentResult result} of the retry.
+     * @param listener The listener object attached with the {@link HttpRequest}.
+     * @param httpRequest The {@link HttpRequest} to send.
+     * @return {@code true} if the request should be retried.
+     */
+    public static boolean shouldRetry(ServerSentResult serverSentResult, ServerSentEventListener listener,
+        HttpRequest httpRequest) {
+        if (Thread.currentThread().isInterrupted() || !listener.shouldRetry(serverSentResult.getException(),
+            serverSentResult.getRetryAfter(), serverSentResult.getLastEventId())) {
+
+            listener.onError(serverSentResult.getException());
+
+            return true;
+        }
 
-            if (retrySSEResult != null && !retryExceptionForSSE(retrySSEResult, listener, httpRequest)
-                && !Thread.currentThread().isInterrupted()) {
-                httpClient.send(httpRequest);
+        if (serverSentResult.getLastEventId() != null) {
+            httpRequest.getHeaders()
+                .add(LAST_EVENT_ID, serverSentResult.getLastEventId());
+        }
+
+        try {
+            if (serverSentResult.getRetryAfter() != null) {
+                Thread.sleep(serverSentResult.getRetryAfter().toMillis());
             }
+        } catch (InterruptedException ignored) {
+            return true;
         }
+
+        return false;
```

Comment:
This method is a little weird. It purports to be a simple boolean getter, but does a lot more than that - log errors to the listener, modify headers, and then do a thread.sleep

Reply:
updated.